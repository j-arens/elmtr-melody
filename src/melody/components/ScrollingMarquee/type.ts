export interface ScrollDimensions {
    overlapTarget: number;
    recordedWidth: number;
}

export interface ScrollStyles {
    transition?: string;
    transform?: string;
}
