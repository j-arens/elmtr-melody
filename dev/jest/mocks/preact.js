module.exports = {
    unmountComponent: jest.fn(),
};
