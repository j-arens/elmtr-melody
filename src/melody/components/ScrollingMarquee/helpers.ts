import { SCROLL_INTERVAL, State } from './';
import { ScrollDimensions, ScrollStyles } from './type';

export enum ScrollStatus {
    LIMIT_LEFT,
    LIMIT_RIGHT,
    IN_RANGE,
}

/**
 * Get the total percentage difference between the
 * offsetWidth and scrollWidth and the elements offsetWidth
 */
export const calculateDimensions = (el: HTMLElement): ScrollDimensions => {
    const offsetWidth = el.offsetWidth;
    const scrollWidth = el.scrollWidth;
    const target = Math.ceil((scrollWidth - offsetWidth) / offsetWidth * 100);
    return {
        overlapTarget: offsetWidth < scrollWidth ? target : 0,
        recordedWidth: offsetWidth,
    };
};

/**
 * Checks if the marquee is in a state where it can start a new scroll
 */
export const canScroll = ({ scrolling, overlapTarget }: State): boolean =>
    Boolean(!scrolling && overlapTarget);

/**
 * Checks if the element resized since the last time the width was recorded
 */
export const didResize = (el, { recordedWidth }: State): boolean =>
    el.offsetWidth !== recordedWidth;

/**
 * Get the current scroll position status
 */
export const getScrollStatus = ({
    offset,
    overlapTarget,
    direction,
}: State): ScrollStatus => {
    if (offset === overlapTarget && direction === 'left') {
        return ScrollStatus.LIMIT_LEFT;
    }
    if (offset === 0 && direction === 'right') {
        return ScrollStatus.LIMIT_RIGHT;
    }
    return ScrollStatus.IN_RANGE;
};

/**
 * Calculate the next offset
 */
export const getNextOffset = ({ offset, direction }: State): number =>
    direction === 'right' ? offset - 1 : offset + 1;

/**
 * Map scroll state into a scroll styles object
 */
export const getScrollStyles = ({
    scrolling,
    overlapTarget,
    offset,
}: State): ScrollStyles => {
    if (!scrolling || !overlapTarget) {
        return {};
    }
    return {
        transform: `translate3d(-${offset}%, 0, 0)`,
        transition: `transform ${SCROLL_INTERVAL}ms linear`,
    };
};
