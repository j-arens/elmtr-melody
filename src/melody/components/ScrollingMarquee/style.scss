.scrollingMarquee {
    &__wrap {
        overflow: hidden;
    }

    &__content {
        display: flex;
        align-items: center;
        white-space: nowrap;
    }
}
