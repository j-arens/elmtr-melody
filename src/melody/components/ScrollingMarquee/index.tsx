import { WithOptionalClassName } from '@components/type';
import { Component, h } from 'preact';
const s = require('./style.scss');

interface Props extends WithOptionalClassName {
    children?: JSX.Element;
}

interface State {
    offset: number;
    target: number;
    scrolling: boolean;
    intervalId: number | NodeJS.Timer;
    direction: string; // left | right
}

interface ScrollStyles {
    transition?: string;
    transform?: string;
}

const SCROLL_INTERVAL = 50;

export default class extends Component<Props, State> {
    state = {
        offset: 0,
        target: 0,
        scrolling: false,
        intervalId: 0,
        direction: 'left',
    };

    componentDidMount() {
        this.calculateOverflow();
    }

    componentDidUpdate(prevProps: Props) {
        if (this.props.children !== prevProps.children) {
            this.reset();
            this.calculateOverflow();
        }
    }

    reset() {
        this.setState({
            offset: 0,
            target: 0,
            scrolling: false,
            intervalId: 0,
            direction: 'left',
        });
    }

    calculateOverflow() {
        const offsetWidth = this.contentRef.offsetWidth;
        const scrollWidth = this.contentRef.scrollWidth;
        if (offsetWidth < scrollWidth) {
            const target = Math.ceil((scrollWidth - offsetWidth) / offsetWidth * 100);
            this.setState({ target });
        }
    }

    startScrolling = () => {
        const { scrolling, target } = this.state;
        if (scrolling || !target) {
            return;
        }
        const intervalId = setInterval(this.updateScroll, SCROLL_INTERVAL);
        this.setState({
            intervalId,
            scrolling: true,
        });
    }

    stopScrolling() {
        const { intervalId } = this.state;
        clearInterval(intervalId);
        this.setState({
            scrolling: false,
            intervalId: 0,
            direction: 'left',
            offset: 0,
        });
    }

    updateScroll = () => {
        const { offset, target, direction } = this.state;
        if (offset === target && direction === 'left') {
            this.setState({ direction: 'right' });
            return;
        }
        if (offset === 0 && direction === 'right') {
            this.stopScrolling();
            return;
        }
        const nextOffset = direction === 'right' ? offset - 1 : offset + 1;
        this.setState({ offset: nextOffset });
    }

    getStyle(): ScrollStyles {
        const { scrolling, target, offset } = this.state;
        if (!scrolling || !target) {
            return {};
        }
        return {
            transform: `translate3d(-${offset}%, 0, 0)`,
            transition: `transform ${SCROLL_INTERVAL}ms linear`,
        };
    }

    setRef = (el: HTMLElement) => {
        this.contentRef = el;
    }

    contentRef: HTMLElement;

    render({ children, className = '' }: Props) {
        return (
            <div
                class={`${s.scrollingMarquee__wrap} ${className}`}
                onMouseEnter={this.startScrolling}
            >
                <div
                    ref={this.setRef}
                    class={s.scrollingMarquee__content}
                    style={this.getStyle()}
                >
                    {children}
                </div>
            </div>
        );
    }
}
