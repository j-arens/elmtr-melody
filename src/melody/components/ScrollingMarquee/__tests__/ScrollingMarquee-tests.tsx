import { h } from 'preact';
import { render } from 'preact-render-to-string';
import ScrollingMarquee from '../';

describe('ScrollingMarquee', () => {
    it('renders', () => {
        const component = render(
            <ScrollingMarquee>
                <div>
                    <p>track</p>
                    <span>-</span>
                    <p>artist</p>
                </div>
            </ScrollingMarquee>,
        );
        expect(component).toMatchSnapshot();
    });
});
