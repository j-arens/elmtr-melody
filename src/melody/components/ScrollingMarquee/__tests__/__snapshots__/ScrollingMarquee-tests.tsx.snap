// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollingMarquee renders 1`] = `
<div class="scrollingMarquee__wrap ">
  <div class="scrollingMarquee__content">
    <div>
      <p>track</p>
      <span>-</span>
      <p>artist</p>
    </div>
  </div>
</div>
`;
